### PR TITLE
API Upgrade

### PR DESCRIPTION
**Pull request by saurabh-khare**
Please refer below ticket for more information on pull request

https://issues.apache.org/jira/browse/SLING-9454

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -43,77 +42,203 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <instructions>
-                        <Export-Package>org.apache.sling.superimposing</Export-Package>
-                        <Private-Package>org.apache.sling.superimposing.impl</Private-Package>
-                        <DynamicImport-Package>org.apache.felix.webconsole</DynamicImport-Package>
-                        <Sling-Nodetypes>SLING-INF/nodetypes/superimposing.cnd</Sling-Nodetypes>
-                    </instructions>
+                    <source>1.8</source>
```
Comment (rombert) [inline pom.xml:74]: Set the `sling.java.version` pom property, you don't need to set this.

File: pom.xml [hunk 2]
```diff
@@ -43,77 +42,203 @@
...
+                        <goals>
+                            <goal>enforce</goal>
+                        </goals>
+                        <configuration>
+                            <rules>
+                                <requireMavenVersion>
+                                    <version>[3.5.4,)</version>
+                                </requireMavenVersion>
+                                <requireJavaVersion>
+                                    <message>Project must be compiled with Java 8 or higher</message>
```
Comment (rombert) [inline pom.xml:61]: If you set the `sling.java.version` property you won't need this.

File: pom.xml [hunk 3]
```diff
@@ -43,77 +42,203 @@
...
-                        <DynamicImport-Package>org.apache.felix.webconsole</DynamicImport-Package>
-                        <Sling-Nodetypes>SLING-INF/nodetypes/superimposing.cnd</Sling-Nodetypes>
-                    </instructions>
+                    <source>1.8</source>
+                    <target>1.8</target>
                 </configuration>
             </plugin>
-
+            <!-- Maven IntelliJ IDEA Plugin -->
+            <plugin>
```
Comment (rombert) [inline pom.xml:79]: You should not need the idea, eclipse and clean plugins declared here.

File: pom.xml [hunk 4]
```diff
@@ -25,8 +25,7 @@
     </parent>
 
     <artifactId>org.apache.sling.superimposing</artifactId>
-    <version>0.2.1-SNAPSHOT</version>
-    <packaging>bundle</packaging>
+    <version>1.0.0</version>
```
Comment (rombert) [inline pom.xml:28]: Please revert the version change, we change to non snapshots only when releasing

File: pom.xml [hunk 5]
```diff
@@ -43,77 +42,203 @@
...
             <plugin>
                 <groupId>org.apache.sling</groupId>
                 <artifactId>maven-sling-plugin</artifactId>
+                <version>2.3.8</version>
+            </plugin>
+            <!-- BND Maven Plugin -->
+            <plugin>
+                <groupId>biz.aQute.bnd</groupId>
+                <artifactId>bnd-maven-plugin</artifactId>
+                <version>5.0.1</version>
```
Comment (rombert) [inline pom.xml:114]: Please remove the version, it's managed in the parent pom. You should upgrade to the latest parent pom while at it ( org.apache.sling:sling-bundle-parent )

File: pom.xml [hunk 6]
```diff
@@ -43,77 +42,203 @@
...
+                <executions>
+                    <execution>
```
Comment (rombert) [inline pom.xml:116]: This execution should come from the parent pom.

File: pom.xml [hunk 7]
```diff
@@ -43,77 +42,203 @@
...
+                        <id>bnd-process</id>
+                        <goals>
+                            <goal>bnd-process</goal>
+                        </goals>
+                        <configuration>
+                            <bnd>
+                                Export-Package: org.apache.sling.superimposing
```
Comment (rombert) [inline pom.xml:123]: We prefer bnd.bnd files instead of pom configurations

File: pom.xml [hunk 8]
```diff
@@ -43,77 +42,203 @@
...
+                                Sling-Nodetypes: SLING-INF/nodetypes/superimposing.cnd
+                            </bnd>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>biz.aQute.bnd</groupId>
+                <artifactId>bnd-baseline-maven-plugin</artifactId>
+                <version>5.0.1</version>
```
Comment (rombert) [inline pom.xml:135]: Please remove this version

File: pom.xml [hunk 9]
```diff
@@ -43,77 +42,203 @@
...
+                <configuration>
+                    <failOnMissing>false</failOnMissing>
+                </configuration>
+                <executions>
```
Comment (rombert) [inline pom.xml:139]: Please remove this execution, it should already be defined in the parent pom.

File: pom.xml [hunk 10]
```diff
@@ -43,77 +42,203 @@
...
+                    <execution>
+                        <id>baseline</id>
+                        <goals>
+                            <goal>baseline</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <!-- Maven Resources Plugin -->
+            <plugin>
```
Comment (rombert) [inline pom.xml:149]: Please remove this plugin declaration

File: pom.xml [hunk 11]
```diff
@@ -43,77 +42,203 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-resources-plugin</artifactId>
+                <version>3.0.2</version>
+            </plugin>
+            <!-- Maven Jar Plugin -->
+            <plugin>
```
Comment (rombert) [inline pom.xml:155]: Please remove this plugin declaration, it will be inherited from the parent pom.

File: pom.xml [hunk 12]
```diff
@@ -43,77 +42,203 @@
...
         </dependency>
         <dependency>
             <groupId>org.apache.sling</groupId>
             <artifactId>org.apache.sling.settings</artifactId>
-            <version>1.0.0</version>
+            <version>1.3.8</version>
             <scope>provided</scope>
         </dependency>
-
+		<dependency>
```
Comment (rombert) [inline pom.xml:187]: Please check the indentation, it seems off compared to the rest of the pom.

File: src/main/java/org/apache/sling/superimposing/SuperimposingManager.java [hunk 13]
```diff
@@ -21,12 +21,28 @@
...
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
 
 /**
  * Manages the resource registrations for the {@link SuperimposingResourceProviderImpl}.
  * Provides read-only access to all registered providers.
  */
 public interface SuperimposingManager {
+	
+	String OBSERVATION_PATHS_PROPERTY = "observationPaths";
```
Comment (rombert) [inline src/main/java/org/apache/sling/superimposing/SuperimposingManager.java:33]: Please check the indentation.

File: src/main/java/org/apache/sling/superimposing/SuperimposingManager.java [hunk 14]
```diff
@@ -38,6 +54,6 @@
      *   Iterator is backed by a {@link java.util.concurrent.ConcurrentHashMap} and is safe to access
      *   even if superimposing resource providers are registered or unregistered at the same time.
      */
-    Iterator<SuperimposingResourceProvider> getRegisteredProviders();
+    Iterator<SuperimposingResourceProviderImpl> getRegisteredProviders();
```
Comment (rombert) [inline src/main/java/org/apache/sling/superimposing/SuperimposingManager.java:57]: This seems incorrect - why are you now returning an implementation type from an interface? Didn't baselining complain?

File: pom.xml (hunk 1 above)
Comment (saurabh-khare) [reply pom.xml]: No longer required, as pom is simplified

File: pom.xml (hunk 2 above)
Comment (saurabh-khare) [reply pom.xml]: No longer required, as pom is simplified

File: pom.xml (hunk 3 above)
Comment (saurabh-khare) [reply pom.xml]: Removed

File: pom.xml (hunk 4 above)
Comment (saurabh-khare) [reply pom.xml]: Reverted to 0.2.1-SNAPSHOT

File: pom.xml (hunk 5 above)
Comment (saurabh-khare) [reply pom.xml]: Done

File: pom.xml (hunk 6 above)
Comment (saurabh-khare) [reply pom.xml]: Done

File: pom.xml (hunk 7 above)
Comment (saurabh-khare) [reply pom.xml]: Added

File: pom.xml (hunk 8 above)
Comment (saurabh-khare) [reply pom.xml]: Removed

File: pom.xml (hunk 9 above)
Comment (saurabh-khare) [reply pom.xml]: Removed

File: pom.xml (hunk 10 above)
Comment (saurabh-khare) [reply pom.xml]: Removed

File: pom.xml (hunk 11 above)
Comment (saurabh-khare) [reply pom.xml]: Removed

File: pom.xml (hunk 12 above)
Comment (saurabh-khare) [reply pom.xml:81]: Fixed

File: src/main/java/org/apache/sling/superimposing/SuperimposingManager.java (hunk 13 above)
Comment (saurabh-khare) [reply src/main/java/org/apache/sling/superimposing/SuperimposingManager.java:33]: Fixed

File: src/main/java/org/apache/sling/superimposing/SuperimposingManager.java (hunk 14 above)
Comment (saurabh-khare) [reply src/main/java/org/apache/sling/superimposing/SuperimposingManager.java]: Collections4 IteratorUtils.unmodifiableIterator does not allow for implicit casting of as compared to Collections IteratorUtils.unmodifiableIterator. 
This is fixed now.

**Comment by saurabh-khare** [comment]
All suggested changes are resolved.

**Comment by rombert** [comment]
Thanks for the update @saurabh-khare . As mentioned in the Jira issue, please use separate commits and individual Jira issues. I'll resume the review then.

**Pull request closed without merging**